### PR TITLE
Implements up-to-date schema tracking

### PR DESCRIPTION
**Pull request by andrejtonev**
SchemaInfo v1
Two distinct implementations: Transactional and Analytical
Supports schema tracking at run-time, replication and durability
Does not support OnDisk

### CODE REVIEW

**Comment by andrejtonev** [comment]
### Tracking
- [ ] **[Link to Epic/Issue]**


### Standard development
- [ ] Update unit/E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch this branch


### CI Testing Labels
- [ ] Select the appropriate CI test labels _(CI -build=**build-name** -test=**test-suite**)_


### Documentation checklist
- [ ] Add the documentation label
- [ ] Add the bug / feature label
- [ ] Add the milestone for which this feature is intended
    - If not known, set for a later milestone
- [ ] Write a release note, including added/changed clauses
    - **[ Release note text ]**
- [ ] **[ Documentation PR link memgraph/documentation#XXXX ]**
    - [ ] Is back linked to this development PR
- [ ] **[ Tag someone from docs team ]**

File: src/storage/v2/inmemory/storage.cpp [hunk 1]
```diff
@@ -2687,6 +2683,52 @@ void InMemoryStorage::CreateSnapshotHandler(
...
+  // a full scan.
+  auto maybe_edge_info = std::invoke([&]() -> EdgeInfo {
+    for (auto &from_vertex : vertices_acc) {
+      auto maybe_edge_info = extract_edge_info(&from_vertex);
+      if (maybe_edge_info) {
+        return maybe_edge_info;
+      }
+    }
+    return std::nullopt;
+  });
```
Comment (Ignition) [inline src/storage/v2/inmemory/storage.cpp:2727]: If we are forced to do a full scan can we do a bulk FindEdges to do less work in total?

File: src/storage/v2/inmemory/storage.cpp [hunk 2]
```diff
@@ -2687,6 +2683,52 @@ void InMemoryStorage::CreateSnapshotHandler(
...
+std::optional<std::tuple<EdgeRef, EdgeTypeId, Vertex *, Vertex *>> InMemoryStorage::FindEdge(Gid gid) {
+  using EdgeInfo = std::optional<std::tuple<EdgeRef, EdgeTypeId, Vertex *, Vertex *>>;
+
+  auto edge_acc = edges_.access();
+  auto edge_it = edge_acc.find(gid);
+  if (edge_it == edge_acc.end()) {
+    return std::nullopt;
+  }
+
+  auto *edge_ptr = &(*edge_it);
```
Comment (Ignition) [inline src/storage/v2/inmemory/storage.cpp:2696]: I assume `DMG_ASSERT(config_.salient.items.properties_on_edges == false);` at this point.

File: src/storage/v2/inmemory/storage.cpp (hunk 1 above)
Comment (Ignition) [reply src/storage/v2/inmemory/storage.cpp:2727]: ignore...too much would need to change.

File: src/storage/v2/delta_container.hpp [hunk 3]
```diff
@@ -16,18 +16,18 @@
...
+using PageAlignedList = std::list<T, utils::PageAlignedAllocator<T>>;
 
 // assumption `sizeof(void *)` if for the node pointer inside forward_list's node
 // multiple pages is also fine, unused pages will not add to RSS
 // using 4 pages (16KiB), because that is the smallest of the large size class in jemalloc
-constexpr auto kRemainingPageSpace = (4 * utils::PageAlignedAllocator<Delta>::PAGE_SIZE) - sizeof(void *);
+constexpr auto kRemainingPageSpace = (4 * utils::PageAlignedAllocator<Delta>::PAGE_SIZE) - 2 * sizeof(void *);
 using delta_slab = memgraph::utils::static_vector<Delta, kRemainingPageSpace>;
 static_assert(alignof(void *) <= alignof(delta_slab), "assumption that above calculation is without any padding");
 static_assert(292 == delta_slab::capacity(),
```
Comment (Ignition) [inline src/storage/v2/delta_container.hpp:33]: no change to capacity....good

File: src/storage/v2/property_store.cpp [hunk 4]
```diff
@@ -756,8 +756,10 @@ std::optional<uint64_t> DecodeZonedTemporalDataSize(Reader &reader) {
     auto tz_str_length = reader.ReadUint(TZ_NAME_LENGTH_SIZE);
     if (!tz_str_length) return std::nullopt;
     zoned_temporal_data_size += (1 + *tz_str_length);
+    reader.SkipBytes(*tz_str_length);
```
Comment (Ignition) [inline src/storage/v2/property_store.cpp:759]: good find

File: src/storage/v2/property_store.cpp [hunk 5]
```diff
@@ -1662,10 +1819,8 @@ auto GetDecodedBuffer(uint8_t (&buffer)[12]) -> DecodedBuffer {
 }
 
 auto GetDecodedBuffer(uint8_t const (&buffer)[12]) -> DecodedBufferConst {
-  uint32_t size = 0;
-  uint8_t *data = nullptr;
-  memcpy(&size, buffer, sizeof(uint32_t));
-  memcpy(&data, buffer + sizeof(uint32_t), sizeof(uint8_t *));
+  uint32_t size = *((uint32_t *)buffer);
+  const uint8_t *data = (uint8_t *)*((uint64_t *)&buffer[sizeof(size)]);
```
Comment (Ignition) [inline src/storage/v2/property_store.cpp:1823]: did this make a difference?

File: src/storage/v2/schema_info.cpp [hunk 6]
```diff
@@ -0,0 +1,854 @@
...
+    if (!vertex_added_) {
+      // Stats have moved; re-read labels pre tx and remove stats
+      tracking_pre_info = &tracking_[PreLabels()];
+      --tracking_pre_info->n;
+
+      // Edges are defined via start/end labels, this identification has changed so update all persistent edges
+      for (const auto &edge : GetRemainingPreEdges<EdgeDirection::IN>(vertex_)) {
+        const auto edge_type = std::get<0>(edge);
+        auto *other_vertex = std::get<1>(edge);
+        const auto edge_ref = std::get<2>(edge);
```
Comment (Ignition) [inline src/storage/v2/schema_info.cpp:497]: structured bindings + we need to change this type away from tuple at some point

File: src/storage/v2/schema_info.hpp [hunk 7]
```diff
@@ -0,0 +1,756 @@
...
+using VertexKey = utils::small_vector<LabelId>;
+
+struct VertexKeyHash {
+  size_t operator()(const utils::small_vector<LabelId> &x) const {
+    uint64_t hash = 0;
+    for (const auto &element : x) {
+      uint32_t val = element.AsUint();
+      val = ((val >> 16) ^ val) * 0x45d9f3b;
+      val = ((val >> 16) ^ val) * 0x45d9f3b;
+      val = (val >> 16) ^ val;
```
Comment (Ignition) [inline src/storage/v2/schema_info.hpp:78]: why not hash_combine?

File: src/storage/v2/schema_info.hpp [hunk 8]
```diff
@@ -0,0 +1,756 @@
...
+  nlohmann::json ToJson(const EnumStore &enum_store, std::string_view key, uint32_t max_count) const;
+};
+struct TrackingInfo {
+  int n{0};
+  std::unordered_map<PropertyId, PropertyInfo> properties;
+
+  nlohmann::json ToJson(NameIdMapper &name_id_mapper, const EnumStore &enum_store) const;
+};
+
+using VertexKey = utils::small_vector<LabelId>;
```
Comment (Ignition) [inline src/storage/v2/schema_info.hpp:69]: How do we ensure the Labels are sorted? (so we have stable hash)

Comment (andrejtonev) [reply src/storage/v2/schema_info.hpp:69]: The hash itself is order independent.
That's also why I don't use hash_combine.

Comment (Ignition) [reply src/storage/v2/schema_info.hpp:69]: Put a comment in for that please.

File: src/storage/v2/property_store.cpp (hunk 5 above)
Comment (andrejtonev) [reply src/storage/v2/property_store.cpp:1823]: No. The performance drop is from a cache miss. Will revert

File: src/storage/v2/inmemory/storage.cpp (hunk 1 above)
Comment (andrejtonev) [reply src/storage/v2/inmemory/storage.cpp:2727]: I just moved the function from the accessor to the storage.
I needed it to enable schema info recovery via WAL files.
v2 I'll add a new delta to WAL (edge set property). But we'll still need the FindEdge for back-compatibility.

File: src/storage/v2/inmemory/storage.cpp (hunk 2 above)
Comment (andrejtonev) [reply src/storage/v2/inmemory/storage.cpp:2696]: You mean `== true`, right?

File: src/storage/v2/property_store.cpp (hunk 5 above)
Comment (andrejtonev) [reply src/storage/v2/property_store.cpp:1823]: Reverted

File: src/storage/v2/delta_container.hpp (hunk 3 above)
Comment (andrejtonev) [reply src/storage/v2/delta_container.hpp:33]: Reverted

**Review by Ignition — APPROVED** [review]

**Pull request merged**